File: day24/haskell/day24.hs
import Control.Arrow ((***), (&&&))
import Data.List.Split (splitOn)
import qualified Data.Set as S

type Component = (Int, Int)
type Bridge = [Component]

parse :: String -> S.Set Component
parse = S.fromList . map (f . splitOn "/") . lines
    where f [i, o] = (read i, read o)

matches :: Int -> S.Set Component -> S.Set Component
matches n = S.filter $ uncurry (||) . ((== n) *** (== n))

construct :: Int -> Bridge -> S.Set Component -> [Bridge]
construct n b cs
    | S.null ms = [b]
    | otherwise = concatMap go ms
    where ms          = matches n cs
          go m@(i, o) = construct (if n == i then o else i) (m:b) (S.delete m cs)

strongest :: [Bridge] -> Int
strongest = maximum . map (sum . map (uncurry (+)))

best :: [Bridge] -> Int
best ps = strongest . filter ((== len) . length) $ ps
    where len = maximum . map length $ ps

solve :: S.Set Component -> (Int, Int)
solve = (strongest &&& best) . construct 0 []

main :: IO ()
main = do
    cs <- parse <$> readFile "day24/input.txt"
    print . solve $ cs

File: day24/typescript/day24.ts
import { readFile } from 'fs';
import { promisify } from 'util';

import * as R from 'ramda';

type Component = [number, number];
type Bridge = Component[];

const parse = (line: string): Component => {
    const [i, o] = R.split('/', line);
    return [+i, +o];
};

const find = (port: number) => R.filter<Component>(
    ([i, o]) => i === port || o === port,
);

const construct = (end: number, cs: Component[], path: Bridge): Bridge[] => {
    const matches = find(end)(cs);
    if (matches.length === 0) {
        return [path];
    }

    return R.chain<Component, Bridge>(
        c => construct(
            c[0] === end ? c[1] : c[0],
            R.remove(cs.indexOf(c), 1, cs),
            [...path, c],
        ),
        matches,
    );
};

const strongest = R.pipe(
    R.map<Bridge, number>(
        R.reduce((acc, [i, o]) => acc + i + o, 0),
    ),
    R.reduce<number, number>(R.max, 0),
);

const solve = (cs: Component[]) => {
    const paths = construct(0, cs, []);

    const part1 = strongest(paths);

    const len = R.reduce<number, number>(R.max, 0, R.map(x => x.length, paths));
    const longest = R.filter(p => p.length === len, paths);
    const part2 = strongest(longest);

    return [part1, part2];
};

(async () => {
    const input = await promisify(readFile)('day24/input.txt', 'utf8');
    const components = R.map(parse, R.split('\r\n', input));

    console.log(solve(components));
})();